README.md
Creatures
=========

Interface with the Creatures 2 game

[![NPM version](http://img.shields.io/npm/v/creatures.svg)](https://npmjs.org/package/creatures) 
[![Flattr this git repo](http://api.flattr.com/button/flattr-badge-large.png)](https://flattr.com/submit/auto?user_id=skerit&url=https://github.com/skerit/creatures&title=Creatures&language=&tags=github&category=software)

## Install

Install it for use in your own application:

`npm install creatures`

## Features

* Executing CAOS commands
* Interpreting history files
* ...

## Author

You can follow me on twitter: [![Twitter](http://i.imgur.com/wWzX9uB.png)@skeriten](https://www.twitter.com/skeriten)

## License

MIT


CHANGELOG.md
## 0.2.8 (WIP)

* Allow maximum 10 entries in the `creatures_queue`, before finally executing the next function in the queue
* Make sure calling back with an error does not block queues
* Add classification system & known scriptorium object names

## 0.2.7 (2018-04-27)

* Remove limitation count on vbole.exe reconnects
* Fix creature update overload
* Only allow 1 egg update script at a time
* Don't move a creature to invalid coordinates
* Make sure local creature info is updated before it is exported
* Retry file reads that fialed with `ENOENT` if they're export files
* Retry getting a parent for generation calculation if the reason is a timeout
* Do not load creature history file if it is empty

## 0.2.6 (2018-04-22)

* Fix getting parents of an egg
* [VB] Also try to identify creatures2.exe against it shortname, creatu~1.exe
* [VB] Add ability to get text content of a window
* [VB] Add `getspeed` command
* Fixed bug where Creature instances not in-game would be destroyed immediately
* Add `CreaturesApplication#creature_count` property, which gives the amount of creatures in the current world
* Cache `CreaturesApplication#getWorldName` for 60 seconds
* Add `Creature#getOwnerData` and `Creature#setNotes`
* Add simple `serialize` and `parse` methods for use in `LString` fields
* [VB] Add retries to getting a window by name or handle, reducing timeouts

## 0.2.5 (2018-03-26)

* [VB] Kill any creatures2.exe process without a window on boot
* [VB] Fix bug where the main "Creatures 2" window was closed
* [VB] No longer iterate through ALL open windows (which was the case when `handle` was 0)
* See what the pause state is before resuming during `doUnpaused`
* Find the correct age of an exported creature

## 0.2.4 (2018-03-26)

* Add `CreaturesApplication#saveGame(callback)` which can save the game when using the `blueberry4$` cheat
* Fix `Creature#teachLanguage()` not working properly
* Add methods to enable powerups
* When no creatures are in the world `getCreatureIds` will no longer callback with an array with 1 empty id string
* Add `log` method to base class
* Get X & Y position of creature on update
* Update creature info before exporting
* You can now supply a position when importing a creature
* Add an initial error dialogbox check on init
* Add `readFile` method that will retry reading a file on `EBUSY` errors
* `Creature#update` can now timeout
* Get the room type a creature is in on update
* Fix gene reading (except dendrites)
* Gene flags are now also parsed
* Implement pigment & pigment bleed
* Add methods to get parents of an egg
* Add methods to get hatchery & egg limit

## 0.2.3 (2018-03-17)

* Make sure `SfcOle` calls the `done` task after receiving error responses
* Find the creatures2 executable in some common places if the game isn't running
* Only run some commands 'till after the ready signal.

## 0.2.2 (2018-03-15)

* Emit `world_name` event when a new world is detected
* Detect removed creatures and emit a `removed` event
* Recreate vbole instance when a new world is opened
* Add command timeout checks
* Fix `Creature#getBodyPartImage` being deadlocked waiting for `ready` event
* Fix getting information for the `World.sfc` world, which stored data in other locations
* If unable to get a gene file in the documents directories, get it from the main application directory
* Fix `S16#load` from getting stuck in an infinite loop
* Fix `Genome#getBodyPartImage` so that it tries all lower lifestages before giving up (especially needed for ettins)
* Get locale & language of the current windows environment
* Add English, Dutch & French menu keys mappings
* Make the Import & Export functions work in all languages (except Japanese)
* Fix the sendKeys functionality on Windows 10 (needed to send {DOWN} to actually open a menu)
* Add some more context info to VBOLE errors
* [VB] Fix identifying other windows with "Creatures 2" as their title as the C2 window (for example explorer windows in a folder named "Creatures 2")
* Add `Creature#select(callback)` method to make it the active creature
* If the creature's name starts with a "<", or it is the moniker or hex moniker, it is deemed "unnamed"
* `CreaturesApplication#getProcessPath` will now callback with the path to the process as a string instead of a wrapped object
* Callback with error if loading history files fails in `Creature#getWorldNames`

## 0.2.1 (2018-03-12)

* VBOle will now always respond with a JSON object than can optionally contain an error
* If a command should fail, other commands in the array will no longer be executed
* [VB] The JSON module in VBOle now looks for a `toJSON` function on class instances
* [VB] VBOle can now get all the child elements of a window
* [VB] VBOle will now look for error dialog boxes and inform the library
* The `SFCOle` class will now emit `vbole_error` errors, for when certain dialogs pop up
* The `CreaturesApplication` class will forward these errors as either `error_dialogbox` or `error_vbole`. A callback is provided with which you have to respond.
* [VB] Add `WriteDebug` method & flush outputs
* [VB] Another VBOle instance is created for monitoring error dialogs during CAOS commands
* [VB] VBOle now accepts a startup argument as json
* Add `SFCOle#escapeKeys` to escape a `SendKeys` string
* Copy `.exp` files to a temporary location before importing them, so they do not get removed
* Add `Export` class to read in exp files (only gets the moniker for now)
* [VB] Get all C2Window child elements, return them to the `c2window_elements` command
* [VB] Add `play` and `pause` command, which sets the "paus" variable using CAOS
* Added `CreaturesApplication` method `play`, `pause`, `getIsPlaying` and `doUnpaused`
* Fix the `Egg` class
* Add `Chemistry` class and chemicals
* Moved `Creature#getBodyPartImage` to `Genome` class
* Add `Creature#inseminate` method

## 0.2.0 (2018-02-17)

* Add speedhack functionality
* Expand vbole capabilities
* Added S16 class

lib/genome.js
var Extractor = require('binary-extractor'),
    Blast = __Protoblast,
    Nr = Blast.Bound.Number,
    Fn = Blast.Collection.Function;

// Get the Creatures namespace
var Creatures = Fn.getNamespace('Develry.Creatures');

/**
 * The Genome class
 *
 * @author   Jelle De Loecker   <jelle@develry.be>
 * @since    0.1.0
 * @version  0.2.0
 *
 * @param    {CreaturesApplication}   app
 * @param    {String}                 path
 */
var Genome = Fn.inherits('Develry.Creatures.Base', function Genome(app, path) {

	// The parent creatures app
	this.app = app;

	// The path to the file
	this.path = path;

	// The genes
	this.genes = [];
});

/**
 * Possible genetypes
 *
 * @author   Jelle De Loecker   <jelle@develry.be>
 * @since    0.1.0
 * @version  0.1.0
 */
Genome.setProperty('types', {
	'00' : {
		name : 'Brain lobe',
		data : [
			// General
			'x',
			'y',
			'width',
			'height',
			'perception',
			// Cell body
			'threshold',
			'leakage',
			'rest_state',
			'input_gain',
			{name: 'state', length: 8},
			'winner_takes_all',

			'__skip__'
			//{name: 'dendrite_0', length: 47},
			//{name: 'dendrite_1', length: 47}
		]
	},
	'01' : {
		name : 'Brain organ',
		data : [
			'clock_rate',
			'life_force_repair_rate',
			'life_force_start',
			'biotick_start',
			'atp_damage_coefficient'
		]
	},
	'10' : {
		name : 'Receptor',
		data : [
			'organ',
			'tissue',
			'locus',
			'chemical',
			'threshold',
			'nominal',
			'gain',
			'flags'
		]
	},
	'11' : {
		name : 'Emitter',
		data : [
			'organ',
			'tissue',
			'locus',
			'chemical',
			'threshold',
			'sample_rate',
			'gain',
			'flags'
		]
	},
	'12' : {
		name : 'Chemical reaction',
		data : [
			'l1a', 'l1c',
			'l2a', 'l2c',
			'r1a', 'r1c',
			'r2a', 'r2c',
			'reaction_rate'
		]
	},
	'13' : {
		name : 'Half lives',
		data : [
			{name: 'values', length: 256}
		]
	},
	'14' : {
		name : 'Initial concentration',
		data : [
			'chemical',
			'amount'
		]
	},
	'20' : {
		name : 'Stimulus',
		data : [
			'type',
			'significance',
			'sensory_neuron',
			'intensity',
			{name: 'flags', values: ['modulate', 'neuron_offset', 'sensed_when_asleep']},
			'chemical_1',
			'chemical_1_amount',
			'chemical_2',
			'chemical_2_amount',
			'chemical_3',
			'chemical_3_amount',
			'chemical_4',
			'chemical_4_amount'
		]
	},
	'21' : {
		name : 'Genus',
		data : [
			{name: 'species', values: ['norn', 'grendel', 'ettin', 'shee']},
			{name: 'mother_moniker', length: 4},
			{name: 'father_moniker', length: 4}
		]
	},
	'22' : {
		name : 'Appearance',
		data : [
			{name: 'body_part', values: ['head', 'body', 'leg', 'arm', 'tail']},
			'breed',
			'species'
		]
	},
	'23' : {
		name : 'Pose',
		data : [
			'pose_number',
			{name: 'pose_string', length: 15}
		]
	},
	'24' : {
		name : 'Gait',
		data : [
			'gait_number',
			'p1', 'p2', 'p3', 'p4', 'p5', 'p6', 'p7', 'p8'
		]
	},
	'25' : {
		name : 'Instinct',
		data : [
			'lobe_nr_1',
			'cell_nr_1',
			'lobe_nr_2',
			'cell_nr_2',
			'lobe_nr_3',
			'cell_nr_3',
			'creature_action',
			'reward_or_punish',
			'amount'
		]
	},
	'26' : {
		name : 'Pigment',
		data : [
			{name: 'color', values: ['red', 'green', 'blue']},
			'intensity'
		]
	},
	'27' : {
		name : 'Pigment bleed',
		data : [
			'rotation',
			'swap'
		]
	},
	'30' : {
		name : 'Organ',
		data : [
			'clock_rate',
			'life_force_repair_rate',
			'life_force_start',
			'biotick_start',
			'atp_damage_coefficient'
		]
	}
});

/**
 * Read the file and process it
 *
 * @author   Jelle De Loecker   <jelle@develry.be>
 * @since    0.1.0
 * @version  0.2.8
 *
 * @param    {Boolean}  refresh
 * @param    {Function} callback
 */
Genome.setMethod(function load(refresh, callback) {

	var that = this;

	if (typeof refresh == 'function') {
		callback = refresh;
		refresh = false;
	}

	if (!callback) {
		callback = Fn.Thrower;
	}

	// Callback if this has already been loaded and a refresh is not needed
	if (this.loaded && !refresh) {
		return Blast.setImmediate(function immediate() {
			callback(null);
		});
	}

	if (!this.path) {
		return callback(new Error('Can not load genome: no path given'));
	}

	this.readFile(this.path, function gotFileBuffer(err, buffer) {

		if (err) {
			return callback(err);
		}

		// Process the buffer
		that.processBuffer(buffer);

		that.loaded = true;

		return callback(null);
	});
});

/**
 * Process the buffer
 *
 * @author   Jelle De Loecker   <jelle@develry.be>
 * @since    0.1.0
 * @version  0.2.6
 *
 * @param    {Buffer|Extractor}   buffer
 * @param    {Boolean}            from_export   Is this buffer from an export file?
 *
 * @return   {Number}   The next index in the buffer that isn't a gene
 */
Genome.setMethod(function processBuffer(buffer, from_export) {

	var gene,
	    temp,
	    gen;

	if (Buffer.isBuffer(buffer)) {
		gen = new Extractor(buffer)
	} else {
		gen = buffer;
		buffer = gen.buffer;
	}

	if (!from_export) {
		// Get the header first
		temp = gen.readBytes(3).toString();

		if (temp != 'dna') {
			throw new Error('Can\'t read file: not a genome');
		}

		// Get the version of the genome file
		this.version = Number(gen.readBytes(1).toString());
	} else {
		// Jump to the first gene part after the CGenome string
		gen.index = buffer.indexOf('gene', buffer.indexOf('CGenome'));
	}

	// Reset the genes array
	this.genes.length = 0;

	// Consume the buffer
	while ((temp = gen.readBytes(4).toString()) == 'gene') {

		// Get the gene header
		gene = {
			// 00, 01, 02 or 03
			type            : gen.readByte(),
			subtype         : gen.readByte(),

			// Specific gene within a type/subtype category
			sequence        : gen.readByte(),

			// Appears to indicate a duplicate gene
			duplicate       : gen.readByte(),

			// The "switch-on" time for the gene
			stage           : gen.readByte(),

			// Contains sex dependence, mutability & dormancy
			flags           : gen.readByte(),
			mutation_chance : gen.readByte()
		};

		// Extract the flags
		gene.mutable     = Nr.bitAt(gene.flags, 0);
		gene.duplicable  = Nr.bitAt(gene.flags, 1);
		gene.deletable   = Nr.bitAt(gene.flags, 2);
		gene.male_only   = Nr.bitAt(gene.flags, 3);
		gene.female_only = Nr.bitAt(gene.flags, 4);
		gene.dormant     = Nr.bitAt(gene.flags, 5);

		gene.male = gene.male_only || gene.female_only == 0;
		gene.female = gene.female_only || gene.male_only == 0;

		// Now get the data itself
		gene.data = this.extractGeneData(gene, gen);

		this.genes.push(gene);
	}

	if (this.genes.length < 50) {
		console.warn('Too few genes found');
	}

	return gen.index - 4;
});

/**
 * Extract gene data
 *
 * @author   Jelle De Loecker   <jelle@develry.be>
 * @since    0.1.0
 * @version  0.2.4
 *
 * @param    {Object}          header
 * @param    {BinaryExtractor} gen
 */
Genome.setMethod(function extractGeneData(header, gen) {

	var typestr = String(header.type) + String(header.subtype),
	    result = {},
	    piece,
	    info,
	    temp,
	    i;

	// Get the name of the type of this gene
	info = this.types[typestr];

	if (!info) {
		this.log('error', 'Unable to find info on gene type', typestr);
		console.info('Header of unknown gene:', header);
		return;
	}

	// Get the name
	header.name = info.name;

	for (i = 0; i < info.data.length; i++) {
		piece = info.data[i];

		// If it's just a string, only 1 byte is needed
		if (typeof piece == 'string') {

			if (piece == '__skip__') {
				temp = gen.after('gene');

				if (temp != -1) {
					gen.index -= 4;
				}
			} else {
				result[piece] = gen.readByte();
			}
		} else {
			if (piece.length) {
				result[piece.name] = gen.readBytes(piece.length);
			} else {
				temp = gen.readByte();
				result[piece.name] = temp;

				// Add the value name
				if (piece.values) {
					result[piece.name + '_name'] = piece.values[temp];
				}
			}
		}
	}

	return result;
});

/**
 * Get all genes of a specific type
 *
 * @author   Jelle De Loecker   <jelle@develry.be>
 * @since    0.2.0
 * @version  0.2.0
 *
 * @param    {String}   name
 *
 * @return   {Array}
 */
Genome.setMethod(function getGenesOfName(name) {

	var result = [],
	    gene,
	    i;

	for (i = 0; i < this.genes.length; i++) {
		gene = this.genes[i];

		if (gene.name == name) {
			result.push(gene);
		}
	}

	return result;
});

/**
 * Get pigment info
 *
 * @author   Jelle De Loecker   <jelle@develry.be>
 * @since    0.2.4
 * @version  0.2.4
 *
 * @return   {Object}
 */
Genome.setMethod(function getPigmentInfo(creature) {

	var intensity,
	    result = {},
	    genes,
	    gene,
	    r,
	    g,
	    b,
	    i;

	// Get the pigment genes
	genes = this.getGenesOfName('Pigment');

	// Prepare the rgb arrays
	r = [];
	g = [];
	b = [];

	for (i = 0; i < genes.length; i++) {
		gene = genes[i];

		if (creature) {
			// Skip genes that are not for this gender
			if ((creature.male && !gene.male) || (creature.female && !gene.female)) {
				continue;
			}
		}

		// Skip dormant genes
		if (gene.dormant) {
			continue;
		}

		if (gene.data && gene.data.color_name) {
			intensity = 1 + (0.125 * ((gene.data.intensity - 128) / 128));

			switch (gene.data.color_name) {
				case 'red':
					r.push(intensity);
					break;

				case 'green':
					g.push(intensity);
					break;

				case 'blue':
					b.push(intensity);
					break;
			}
		}
	}

	if (r.length) {
		r = Creatures.S16.meanReduce(r);
	} else {
		r = 1;
	}

	if (g.length) {
		g = Creatures.S16.meanReduce(g);
	} else {
		g = 1;
	}

	if (b.length) {
		b = Creatures.S16.meanReduce(b);
	} else {
		b = 1;
	}

	result = {
		r : r,
		g : g,
		b : b
	};

	return result;
});

/**
 * Get pigment bleed info
 *
 * @author   Jelle De Loecker   <jelle@develry.be>
 * @since    0.2.4
 * @version  0.2.4
 *
 * @return   {Object}
 */
Genome.setMethod(function getPigmentBleedInfo(creature) {

	var result = [],
	    rotations = [],
	    swaps = [],
	    genes,
	    gene,
	    i;

	// Get the pigment genes
	genes = this.getGenesOfName('Pigment bleed');

	for (i = 0; i < genes.length; i++) {
		gene = genes[i];

		if (creature) {
			// Skip genes that are not for this gender
			if ((creature.male && !gene.male) || (creature.female && !gene.female)) {
				continue;
			}
		}

		// Skip dormant genes
		if (gene.dormant) {
			continue;
		}

		if (gene.data) {
			result.push(gene.data);

			// Calculate the rotation coefficient
			gene.data.rotation_coef = (0.33 * ((gene.data.rotation - 128) / 128));

			if (gene.data.rotation >= 128) {
				gene.data.abs_rot = gene.data.rotation - 128;
			} else {
				gene.data.abs_rot = 128 - gene.data.rotation;
			}

			gene.data.inv_rot = 127 - gene.data.abs_rot;

			// Calculate the swap coefficient
			gene.data.swap_coef = Math.abs(1 * ((gene.data.swap - 128) / 128));

			if (gene.data.swap >= 128) {
				gene.data.abs_swap = gene.data.swap - 128;
			} else {
				gene.data.abs_swap = 128 - gene.data.swap;
			}

			gene.data.inv_swap = 127 - gene.data.abs_swap;
		}
	}

	return result;
});


/**
 * Get a specific body image of this creature
 *
 * @author   Jelle De Loecker   <jelle@develry.be>
 * @since    0.2.0
 * @version  0.2.4
 *
 * @param    {Export|Creature}   creature
 * @param    {String}            part_name    The body part type
 * @param    {Function}          callback
 *
 * @return   {Creatures.S16}
 */
Genome.setMethod(function getBodyPartImage(creature, part_name, callback) {

	var that = this,
	    breed_char,
	    filename,
	    try_age,
	    prefix,
	    parts,
	    genes,
	    gene,
	    s16,
	    i;

	parts = {
		head             : 'A',
		body             : 'B',
		left_thigh       : 'C',
		left_shin        : 'D',
		left_foot        : 'E',
		right_thigh      : 'F',
		right_shin       : 'G',
		right_foot       : 'H',

		left_humerus     : 'I',
		left_upper_arm   : 'I',

		left_radius      : 'J',
		left_forearm     : 'J',

		right_radius     : 'K',
		right_forearm    : 'K',

		// Tails are not in C1
		tail_root        : 'M',
		tail_tip         : 'N',

		// Ears & hair are present in CV
		// but supported in DS
		left_ear         : 'O',
		right_ear        : 'P',
		hair             : 'Q'
	};

	// The first char of the filename is the body part
	prefix = parts[part_name];

	// Get the appearance genes
	genes = that.getGenesOfName('Appearance');

	for (i = 0; i < genes.length; i++) {
		gene = genes[i];

		if (gene.data.body_part_name == part_name) {
			break;
		} else {
			gene = null
		}
	}

	if (!gene) {
		console.warn('Could not find', part_name, 'gene in', genes, this);

		return callback(null);
	}

	// Add the species number
	if (creature.female) {
		prefix += (gene.data.species + 4);
	} else {
		prefix += gene.data.species;
	}

	// Not all lifestages are available,
	// try with the current one first
	try_age = creature.agen || 0;

	// The last part of the filename is the breed character
	breed_char = String.fromCharCode(65 + gene.data.breed);

	// The filename we want
	filename = prefix + try_age + breed_char + '.s16';

	// Create the s16 instance
	s16 = new Creatures.S16(that.app, filename);

	// Set the pigment info
	s16.setPigment(this.getPigmentInfo(creature));
	s16.setPigmentBleed(this.getPigmentBleedInfo(creature));

	s16.load(function loaded(err) {

		if (err) {

			if (err.code == 'ENOENT' && try_age > 0) {
				try_age--;

				// Construct the new filename of a lower lifestage
				// (Ettins for example only have sprites for lifestage 0)
				filename = prefix + try_age + breed_char + '.s16';

				s16 = new Creatures.S16(that.app, filename);

				return s16.load(loaded);
			}

			return callback(err);
		}

		callback(null, s16);
	});

	return s16;
});

module.exports = Genome;